go/libraries/doltcore/sqle/database_provider.go

```diff
@@ -1007,20 +1007,103 @@ func isBranch(ctx context.Context, db SqlDatabase, branchName string, dialer dbf
 		return "", false, fmt.Errorf("unrecognized type of database %T", db)
 	}
 
+	brName, branchExists, err := isLocalBranch(ctx, ddbs, branchName)
+	if err != nil {
+		return "", false, err
+	}
+	if branchExists {
+		return brName, true, nil
+	}
+
+	brName, branchExists, err = isRemoteBranch(ctx, db, ddbs, branchName)
+	if err != nil {
+		return "", false, err
+	}
+	if branchExists {
+		return brName, true, nil
+	}
+
+	return "", false, nil
+}
+
+func isLocalBranch(ctx context.Context, ddbs []*doltdb.DoltDB, branchName string) (string, bool, error) {
 	for _, ddb := range ddbs {
-		branchName, branchExists, err := ddb.HasBranch(ctx, branchName)
+		brName, branchExists, err := ddb.HasBranch(ctx, branchName)
 		if err != nil {
 			return "", false, err
 		}
 
 		if branchExists {
-			return branchName, true, nil
+			return brName, true, nil
 		}
 	}
 
 	return "", false, nil
 }
 
+// isRemoteBranch is called only when the branch in connection string does not exist locally. It searches for a remote branch.
+// If there is a match, it creates a new local branch from the remote branch and sets its upstream to the remote branch.
+func isRemoteBranch(ctx context.Context, srcDB SqlDatabase, ddbs []*doltdb.DoltDB, branchName string) (string, bool, error) {
```

Comment:
"only when the branch in connection string does not exist locally" probably isn't totally accurate, since I think this behavior would also kick in when someone does `use mydb/a-remote-tracking-branch`. 

Would be good to use the term "remote tracking branch" to disambiguate between a "remote branch" and "remote tracking branch". It's a small distinction, but I think an important one – "remote tracking branch" implies that we have a local tracking branch in our db metadata that tells us about the remote branch. 

From just the name of this function, I wasn't expecting it to create the local branch from the remote tracking branch (although, I do see you called that out in the function docs). I think this would be a little more intuitive to future readers if you either separated the existence check from the creation of a local branch from the remote tracking branch. Alternatively, you could also consider changing the name of this function to something like `createLocalBranchFromRemoteTrackingBranch` or something else nice and self-descriptive.